File: dashpilot-android/app/src/main/java/com/softwiredtech/dashpilot/datasource/DashKitOtaUpdate.kt
package com.softwiredtech.dashpilot.datasource

import android.annotation.SuppressLint
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.content.Context
import android.os.Build
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.UUID

sealed class OtaState {
    object Idle : OtaState()
    object Connecting : OtaState()
    data class Uploading(val progress: Float) : OtaState()
    object Rebooting : OtaState()
    data class Error(val message: String) : OtaState()
}

@SuppressLint("MissingPermission")
class DashKitOtaUpdate(
    private val manager: DashKitBleManager,
    private val context: Context
) : GattListener {

    companion object {
        private const val TAG = "DashKitOta"
        private val OTA_SERVICE_UUID = UUID.fromString("CADA0100-CA00-B1E0-B0D6-C000AA0100A1")
        private val OTA_CTRL_UUID = UUID.fromString("CADA0101-CA00-B1E0-B0D6-C000AA0100A1")
        private val OTA_DATA_UUID = UUID.fromString("CADA0102-CA00-B1E0-B0D6-C000AA0100A1")
        private val OTA_STATUS_UUID = UUID.fromString("CADA0103-CA00-B1E0-B0D6-C000AA0100A1")
        private val CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
        private const val FIRMWARE_ASSET = "dashkit-firmware.bin"
    }

    private val _state = MutableStateFlow<OtaState>(OtaState.Idle)
    val state: StateFlow<OtaState> = _state

    private var firmware: ByteArray? = null
    private var firmwareOffset = 0
    private var ctrlChar: BluetoothGattCharacteristic? = null
    private var dataChar: BluetoothGattCharacteristic? = null
    private var statusChar: BluetoothGattCharacteristic? = null

    fun start() {
        val fw = loadFirmware()
        if (fw == null || fw.isEmpty()) {
            _state.value = OtaState.Error("Firmware file not found or empty")
            return
        }
        firmware = fw
        firmwareOffset = 0

        manager.addGattListener(this)

        // If already connected, use existing GATT directly
        val g = manager.gatt
        if (g != null && manager.connectionState.value == ConnectionStatus.Connected) {
            setupOtaService(g)
        } else {
            _state.value = OtaState.Connecting
            manager.connect()
        }
    }

    fun cancel() {
        manager.removeGattListener(this)
        firmware = null
        ctrlChar = null
        dataChar = null
        statusChar = null
        _state.value = OtaState.Idle
    }

    override fun onServicesReady(gatt: BluetoothGatt) {
        setupOtaService(gatt)
    }

    private fun setupOtaService(g: BluetoothGatt) {
        val service = g.getService(OTA_SERVICE_UUID)
        if (service == null) {
            _state.value = OtaState.Error("OTA service not found on device")
            return
        }

        ctrlChar = service.getCharacteristic(OTA_CTRL_UUID)
        dataChar = service.getCharacteristic(OTA_DATA_UUID)
        statusChar = service.getCharacteristic(OTA_STATUS_UUID)

        if (ctrlChar == null || dataChar == null || statusChar == null) {
            _state.value = OtaState.Error("OTA characteristics not found")
            return
        }

        g.setCharacteristicNotification(statusChar, true)
        val descriptor = statusChar!!.getDescriptor(CCCD_UUID)
        if (descriptor != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                g.writeDescriptor(descriptor, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
            } else {
                @Suppress("DEPRECATION")
                descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                @Suppress("DEPRECATION")
                g.writeDescriptor(descriptor)
            }
        }
        Log.d(TAG, "OTA service ready, subscribed to status")
    }

    override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
        // Only handle OTA status descriptor
        if (descriptor.characteristic?.uuid != OTA_STATUS_UUID) return
        if (status != BluetoothGatt.GATT_SUCCESS) {
            _state.value = OtaState.Error("Failed to enable OTA notifications")
            return
        }
        sendBeginCommand(gatt)
    }

    override fun onCharacteristicChanged(
        gatt: BluetoothGatt,
        characteristic: BluetoothGattCharacteristic,
        value: ByteArray
    ) {
        if (characteristic.uuid != OTA_STATUS_UUID) return
        handleStatusNotification(value)
    }

    override fun onCharacteristicWrite(
        gatt: BluetoothGatt,
        characteristic: BluetoothGattCharacteristic,
        status: Int
    ) {
        if (characteristic.uuid != OTA_CTRL_UUID && characteristic.uuid != OTA_DATA_UUID) return
        if (status != BluetoothGatt.GATT_SUCCESS) {
            _state.value = OtaState.Error("Write failed (status $status)")
            return
        }
        sendNextChunk(gatt)
    }

    override fun onDisconnected() {
        val currentState = _state.value
        if (currentState !is OtaState.Rebooting && currentState !is OtaState.Idle) {
            _state.value = OtaState.Error("Disconnected unexpectedly")
        }
        firmware = null
        ctrlChar = null
        dataChar = null
        statusChar = null
    }

    private fun loadFirmware(): ByteArray? {
        return try {
            context.assets.open(FIRMWARE_ASSET).use { it.readBytes() }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load firmware: ${e.message}")
            null
        }
    }

    private fun sendBeginCommand(g: BluetoothGatt) {
        val fw = firmware ?: return
        val size = fw.size
        val cmd = byteArrayOf(
            0x01,
            (size and 0xFF).toByte(),
            ((size shr 8) and 0xFF).toByte(),
            ((size shr 16) and 0xFF).toByte(),
            ((size shr 24) and 0xFF).toByte()
        )
        Log.d(TAG, "Sending OTA Begin: ${fw.size} bytes")
        _state.value = OtaState.Uploading(0f)
        firmwareOffset = 0

        val ctrl = ctrlChar ?: return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            g.writeCharacteristic(ctrl, cmd, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT)
        } else {
            @Suppress("DEPRECATION")
            ctrl.value = cmd
            ctrl.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            @Suppress("DEPRECATION")
            g.writeCharacteristic(ctrl)
        }
    }

    private fun sendNextChunk(g: BluetoothGatt) {
        val fw = firmware ?: return
        if (firmwareOffset >= fw.size) return

        val chunkSize = minOf(manager.mtu - 3, fw.size - firmwareOffset)
        val chunk = fw.copyOfRange(firmwareOffset, firmwareOffset + chunkSize)
        firmwareOffset += chunkSize

        val progress = firmwareOffset.toFloat() / fw.size.toFloat()
        _state.value = OtaState.Uploading(progress)

        val data = dataChar ?: return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            g.writeCharacteristic(data, chunk, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT)
        } else {
            @Suppress("DEPRECATION")
            data.value = chunk
            data.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            @Suppress("DEPRECATION")
            g.writeCharacteristic(data)
        }
    }

    private fun handleStatusNotification(value: ByteArray) {
        if (value.isEmpty()) return
        when (value[0].toInt() and 0xFF) {
            0x01 -> {
                if (value.size >= 5) {
                    val received = (value[1].toInt() and 0xFF) or
                            ((value[2].toInt() and 0xFF) shl 8) or
                            ((value[3].toInt() and 0xFF) shl 16) or
                            ((value[4].toInt() and 0xFF) shl 24)
                    Log.d(TAG, "Device ack: $received bytes received")
                }
            }
            0x02 -> {
                Log.d(TAG, "OTA complete, device rebooting")
                _state.value = OtaState.Rebooting
                manager.removeGattListener(this)
                firmware = null
            }
            0xFF -> {
                val errCode = if (value.size > 1) value[1].toInt() and 0xFF else 0
                Log.e(TAG, "OTA error from device: 0x${errCode.toString(16)}")
                _state.value = OtaState.Error("Device reported error (0x${errCode.toString(16)})")
                manager.removeGattListener(this)
                firmware = null
            }
        }
    }
}

File: dashpilot-ios/dashpilot/BLE/DashKitOtaUpdate.swift
import Foundation
import CoreBluetooth
import Observation

enum OtaState: Equatable {
    case idle
    case connecting
    case uploading(Float)
    case rebooting
    case error(String)
}

/// BLE firmware upload to the DashKit's OTA service (port of the Android
/// `DashKitOtaUpdate`). Protocol:
///   1. enable notifications on the status characteristic (CADA0103),
///   2. write Begin [0x01][size LE32] to the control characteristic (CADA0101),
///   3. stream the binary in write-with-response chunks to CADA0102, paced by
///      the write acks,
///   4. status notifications report acks (0x01), completion/reboot (0x02) or a
///      firmware error (0xFF).
@Observable
final class DashKitOtaUpdate: DashKitGattListener {

    private(set) var state: OtaState = .idle

    @ObservationIgnored
    var onCompleted: (() -> Void)?

    private let manager: DashKitBleManager

    // Mutated only on the manager's BLE queue (listener callbacks) once the
    // upload is running.
    private var firmware: Data?
    private var firmwareOffset = 0
    private var peripheral: CBPeripheral?
    private var ctrlChar: CBCharacteristic?
    private var dataChar: CBCharacteristic?
    private var statusChar: CBCharacteristic?

    // The CAN stream is paused for the upload — its high-rate notifications
    // compete with the OTA writes for connection-event bandwidth. Resumed on
    // error/cancel; after a successful upload the DashKit reboots and the
    // data source re-subscribes on reconnect.
    private var canChar: CBCharacteristic?
    private var canWasNotifying = false

    private static let pingInterval: TimeInterval = 15
    private var controlChar: CBCharacteristic?
    private var lastPingAt = Date.distantPast
    private var pingInFlight = false

    // Set synchronously on the BLE queue when the device reports completion;
    // `state` is published on main so it can lag the disconnect that follows.
    private var rebooting = false
    private var uploadFinished = false

    init(manager: DashKitBleManager) {
        self.manager = manager
    }

    func start(firmware fw: Data) {
        guard !fw.isEmpty else {
            setState(.error("Firmware file is empty"))
            return
        }
        firmware = fw
        firmwareOffset = 0
        rebooting = false
        uploadFinished = false
        manager.suppressPings = true
        setState(.connecting)
        // If already connected the manager replays onServicesReady right away;
        // otherwise kick off a connection.
        manager.addGattListener(self)
        manager.connect()
    }

    func cancel() {
        manager.removeGattListener(self)
        manager.suppressPings = false
        rebooting = false
        uploadFinished = false
        resumeCanNotifications()
        firmware = nil
        peripheral = nil
        ctrlChar = nil
        dataChar = nil
        statusChar = nil
        canChar = nil
        controlChar = nil
        pingInFlight = false
        setState(.idle)
    }

    private func resumeCanNotifications() {
        if canWasNotifying, let peripheral, let canChar {
            peripheral.setNotifyValue(true, for: canChar)
        }
        canWasNotifying = false
    }

    private func setState(_ newState: OtaState) {
        DispatchQueue.main.async { self.state = newState }
    }

    // MARK: - DashKitGattListener (called on the manager's BLE queue)

    func onServicesReady(_ peripheral: CBPeripheral) {
        if rebooting {
            // The DashKit came back on the new firmware: the update is done.
            rebooting = false
            manager.removeGattListener(self)
            setState(.idle)
            onCompleted?()
            return
        }
        guard let service = peripheral.services?.first(where: { $0.uuid == DashKitGatt.otaService }) else {
            setState(.error("OTA service not found on device"))
            return
        }
        self.peripheral = peripheral
        ctrlChar = service.characteristics?.first { $0.uuid == DashKitGatt.otaControlCharacteristic }
        dataChar = service.characteristics?.first { $0.uuid == DashKitGatt.otaDataCharacteristic }
        statusChar = service.characteristics?.first { $0.uuid == DashKitGatt.otaStatusCharacteristic }
        guard let statusChar, ctrlChar != nil, dataChar != nil else {
            setState(.error("OTA characteristics not found"))
            return
        }
        let canService = peripheral.services?.first { $0.uuid == DashKitGatt.canService }
        canChar = canService?.characteristics?.first { $0.uuid == DashKitGatt.canCharacteristic }
        controlChar = canService?.characteristics?.first { $0.uuid == DashKitGatt.controlCharacteristic }
        if let canChar, canChar.isNotifying {
            canWasNotifying = true
            peripheral.setNotifyValue(false, for: canChar)
            print("[DashKitOta] paused CAN notifications for the upload")
        }
        peripheral.setNotifyValue(true, for: statusChar)
        print("[DashKitOta] OTA service ready, subscribing to status")
    }

    func onNotificationStateUpdated(_ characteristic: CBCharacteristic, error: Error?) {
        guard characteristic.uuid == DashKitGatt.otaStatusCharacteristic else { return }
        if error != nil {
            setState(.error("Failed to enable OTA notifications"))
            resumeCanNotifications()
            return
        }
        sendBeginCommand()
    }

    func onCharacteristicChanged(_ characteristic: CBCharacteristic, value: Data) {
        guard characteristic.uuid == DashKitGatt.otaStatusCharacteristic else { return }
        handleStatusNotification(value)
    }

    func onCharacteristicWrite(_ characteristic: CBCharacteristic, error: Error?) {
        if characteristic.uuid == DashKitGatt.controlCharacteristic {
            guard pingInFlight else { return }
            pingInFlight = false
        } else if characteristic.uuid != DashKitGatt.otaControlCharacteristic,
                  characteristic.uuid != DashKitGatt.otaDataCharacteristic {
            return
        }
        if let error {
            setState(.error("Write failed (\(error.localizedDescription))"))
            resumeCanNotifications()
            return
        }
        if characteristic.uuid == DashKitGatt.otaDataCharacteristic,
           let fw = firmware, firmwareOffset >= fw.count {
            uploadFinished = true
        }
        sendNextChunk()
    }

    func onDisconnected() {
        // Rebooting expects this drop; the listener stays registered so the
        // reconnect's onServicesReady can clear the completed state.
        if uploadFinished, !rebooting {
            print("[DashKitOta] link dropped after the last chunk; treating as reboot")
            rebooting = true
            setState(.rebooting)
        } else if !rebooting, state != .idle {
            setState(.error("Disconnected unexpectedly"))
        }
        manager.suppressPings = false
        firmware = nil
        peripheral = nil
        ctrlChar = nil
        dataChar = nil
        statusChar = nil
        canChar = nil
        controlChar = nil
        canWasNotifying = false
        pingInFlight = false
    }

    // MARK: - Upload

    private func sendBeginCommand() {
        guard let fw = firmware, let peripheral, let ctrl = ctrlChar else { return }
        let size = fw.count
        let cmd = Data([
            0x01,
            UInt8(size & 0xFF),
            UInt8((size >> 8) & 0xFF),
            UInt8((size >> 16) & 0xFF),
            UInt8((size >> 24) & 0xFF)
        ])
        print("[DashKitOta] sending OTA Begin: \(size) bytes")
        setState(.uploading(0))
        firmwareOffset = 0
        lastPingAt = .distantPast
        pingInFlight = false
        peripheral.writeValue(cmd, for: ctrl, type: .withResponse)
    }

    private func sendNextChunk() {
        guard let fw = firmware, let peripheral, let data = dataChar else { return }
        guard firmwareOffset < fw.count else { return }

        if let controlChar, Date().timeIntervalSince(lastPingAt) >= Self.pingInterval {
            lastPingAt = Date()
            pingInFlight = true
            let ping = VehicleControl.payload(opcode: VehicleControl.cmdPing, value: 1)
            peripheral.writeValue(ping, for: controlChar, type: .withResponse)
            return
        }

        // .withoutResponse reports the true MTU-3 payload; .withResponse
        // reports 512, which turns each chunk into a slow ATT long write.
        let maxLen = peripheral.maximumWriteValueLength(for: .withoutResponse)
        let chunkSize = min(maxLen, fw.count - firmwareOffset)
        let chunk = fw.subdata(in: firmwareOffset..<(firmwareOffset + chunkSize))
        firmwareOffset += chunkSize

        setState(.uploading(Float(firmwareOffset) / Float(fw.count)))
        peripheral.writeValue(chunk, for: data, type: .withResponse)
    }

    private func handleStatusNotification(_ value: Data) {
        guard let first = value.first else { return }
        switch first {
        case 0x01:
            if value.count >= 5 {
                let received = UInt32(value[1])
                    | (UInt32(value[2]) << 8)
                    | (UInt32(value[3]) << 16)
                    | (UInt32(value[4]) << 24)
                print("[DashKitOta] device ack: \(received) bytes received")
            }
        case 0x02:
            print("[DashKitOta] OTA complete, device rebooting")
            rebooting = true
            setState(.rebooting)
            manager.suppressPings = false
            firmware = nil
        case 0xFF:
            let errCode = value.count > 1 ? value[1] : 0
            print("[DashKitOta] OTA error from device: 0x\(String(errCode, radix: 16))")
            setState(.error("Device reported error (0x\(String(errCode, radix: 16)))"))
            manager.removeGattListener(self)
            manager.suppressPings = false
            uploadFinished = false
            resumeCanNotifications()
            firmware = nil
        default:
            break
        }
    }
}
